### PR TITLE
Fix fn_sig_for_fn_abi and the coroutine transform for generators

### PR DESCRIPTION
**Pull request by bjorn3**
There were three issues previously:
* The self argument was pinned, despite Iterator::next taking an unpinned mutable reference.
* A resume argument was passed, despite Iterator::next not having one.
* The return value was CoroutineState<Item, ()> rather than Option<Item>

While these things just so happened to work with the LLVM backend, cg_clif does much stricter checks when trying to assign a value to a place. In addition it can't handle the mismatch between the amount of arguments specified by the FnAbi and the FnSig.

### CODE REVIEW

**Comment by bjorn3** [comment]
Fixes https://github.com/rust-lang/rust/pull/116447#issuecomment-1790451463

File: compiler/rustc_ty_utils/src/abi.rs [hunk 1]
```diff
@@ -112,7 +112,13 @@ fn fn_sig_for_fn_abi<'tcx>(
             let pin_did = tcx.require_lang_item(LangItem::Pin, None);
             let pin_adt_ref = tcx.adt_def(pin_did);
             let pin_args = tcx.mk_args(&[env_ty.into()]);
-            let env_ty = Ty::new_adt(tcx, pin_adt_ref, pin_args);
+            let env_ty = if tcx.coroutine_is_gen(did) {
```
Comment (compiler-errors) [inline compiler/rustc_ty_utils/src/abi.rs:115]: I wonder if we should just exhaustively match on the coroutine kind

File: compiler/rustc_ty_utils/src/abi.rs [hunk 2]
```diff
@@ -121,6 +127,8 @@ fn fn_sig_for_fn_abi<'tcx>(
             // function in case this is a special coroutine backing an async construct.
             let (resume_ty, ret_ty) = if tcx.coroutine_is_async(did) {
```
Comment (compiler-errors) [inline compiler/rustc_ty_utils/src/abi.rs:128]: here too could be exhaustive

**Review by compiler-errors — APPROVED** [review]
r=me with or without doing more exhaustive matching on coroutine kind

**Comment by compiler-errors** [comment]
Thanks @bjorn3 for catching this. I really appreciate that clif is a lot less forgiving with ABI -- you've already caught a few of these mismatched in the past.

File: compiler/rustc_ty_utils/src/abi.rs (hunk 1 above)
Comment (bjorn3) [reply compiler/rustc_ty_utils/src/abi.rs]: Done

File: compiler/rustc_ty_utils/src/abi.rs (hunk 2 above)
Comment (bjorn3) [reply compiler/rustc_ty_utils/src/abi.rs]: And done.

**Comment by bjorn3** [comment]
Added exhaustive matching and fixed the assertion on the resume type.

**Comment by dlight** [comment]
> While these things just so happened to work with the LLVM backend, cg_clif does much stricter checks when trying to assign a value to a place.

[Here](https://github.com/rust-lang/rust/pull/116447#issuecomment-1790451463) you did mention that despite the LLVM backend accepting it, it was triggering UB. Does this PR fixes this for LLVM as well?

**Comment by bjorn3** [comment]
Yes

**Pull request merged**